internal/telemetry/message.go

```diff
@@ -0,0 +1,103 @@
+	RequestTypeAppClosing      RequestType = "app-closing"
+)
+
+// Application is identifying information about the app itself
+type Application struct {
+	ServiceName     string `json:"service_name"`
+	Env             string `json:"env"`
+	ServiceVersion  string `json:"service_version,omitempty"`
+	TracerVersion   string `json:"tracer_version"`
+	LanguageName    string `json:"language_name"`
+	LanguageVersion string `json:"language_version"`
+}
+
+// Host is identifying information about the host on which the app
+// is running
+type Host struct {
+	ContainerID string `json:"container_id,omitempty"`
+	Hostname    string `json:"hostname,omitempty"`
+	OS          string `json:"os,omitempty"`
+	OSVersion   string `json:"os_version,omitempty"`
+	// TODO: Do we care about the kernel stuff? internal/osinfo gets most of
+	// this information in OSName/OSVersion
+	KernelName    string `json:"kernel_name,omitempty"`
+	KernelRelease string `json:"kernel_release,omitempty"`
+	KernelVersion string `json:"kernel_version,omitempty"`
+}
+
+// AppStarted corresponds to the "app-started" request type
+type AppStarted struct {
+	Integrations  []Integration   `json:"integrations"`
+	Dependencies  []Dependency    `json:"dependencies"`
+	Configuration []Configuration `json:"configuration"`
+}
+
+type Integration struct {
+	Name        string `json:"name"`
+	Enabled     bool   `json:"enabled"`
+	Version     string `json:"version,omitempty"`
+	AutoEnabled bool   `json:"auto_enabled,omitempty"`
+	Compatible  bool   `json:"compatible,omitempty"`
+	Error       string `json:"error,omitempty"`
+}
+
+// Dependency is a Go module on which the applciation depends. This information
+// can be accesed at run-time through the runtime/debug.ReadBuildInfo API.
+type Dependency struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+}
+
+// Configuration is a library-specific configuration value
+type Configuration struct {
+	Name string `json:"name"`
+	// Value should have a type that can be marshaled to JSON
+	Value interface{} `json:"value"`
+}
+
+// Metrics corresponds to the "generate-metrics" request type
+type Metrics struct {
+	Namespace   string   `json:"namespace"`
+	LibLanguage string   `json:"lib_language"`
+	LibVersion  string   `json:"lib_version"`
+	Series      []Series `json:"series"`
+}
+
+// Series is a sequence of observations for a single named metric
+type Series struct {
+	Name   string       `json:"name"`
+	Points [][2]float64 `json:"points"`
+	Type   string       `json:"type"`
+	Tags   []string     `json:"tags,omitempty"`
+	// Common distinguishes metrics which are cross-language vs.
+	// language-specific.
+	//
+	// NOTE: If this field isn't present in the request, the API assumes
+	// assumed the metric is common. So we can't "omitempty" even though the
+	// field is technically optional.
+	Common bool `json:"common"`
+}
+
+// TODO: app-dependencies-loaded and app-integrations-change? Does this really
+// apply to Go?
```

Comment:
AppSec will for vuln management :-)
But happy to do it once we need it in the next few months.

Reply:
Also, it's been useful to debug our user setups by being able to see what modules do they use and with which version.

Reply:
For now I do use `runtime.ReadBuildInfo` to send the modules and versions in the "app-started" message. As far as I know Go doesn't support dynamic loading of dependencies after the app started, right?

Reply:
Ok, indeed, it's very unlikely that we'll have to monitor dl_opens ^^
👍